### ISSUE TITLE
kubectl taint not backwards-compatible to 1.5.x clusters

### ISSUE THREAD

**Issue by skriss** [kubernetes/kubernetes#42614]
@davidopp please give me your take on this issue

In the following skew test running kubectl 1.6.x against a 1.5.x cluster, the kubectl taint tests are failing (update taint and remove taint in particular). Is this expected/acknowledged as an exception to the version skew policy?

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-1.5-1.6-gci-kubectl-skew/3

**Comment by aveshagarwal** [comment]
Yes its expected. See here https://groups.google.com/forum/#!topic/kubernetes-dev/-fwfHQtm9GE

**Comment by davidopp** [comment]
Yeah, as @aveshagarwal alluded to, the 1.6 kubectl will only work against a 1.6 master, and vice versa. (For `kubectl taint`.)

**Comment by skriss** [comment]
@aveshagarwal @davidopp thanks! I have a number of other skew test failures related to Affinity/Taints that I'd like you to just double-check and ensure they're non-issues. An example job output is here:

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gke-container_vm-1.5-gci-1.6-upgrade-master/18

This job creates a 1.5 cluster, upgrades the master to 1.6, and then runs the 1.5 E2E test suite against it to test for backwards compatibility.

Can you confirm that any of the failing tests starting with **SchedulerPredicates** are expected failures? Thanks!

**Comment by skriss** [comment]
@aveshagarwal @davidopp ping on this - would like to make sure these are non-issues else remediate ASAP.

**Comment by davidopp** [comment]
All of the 1.5-1.6 and 1.6-1.5 skew tests are currently passing except the two that I would expect to fail due to backward-incompatible change in "kubectl taint" behavior. One of the fixes in the past few days must have fixed it. So I'm going to close this issue; please reopen if there is still a problem.

**Comment by skriss** [comment]
@davidopp there are still some failures in the upgrade test jobs - I believe they are non-issues given the breaking changes you described, but need you to confirm. See the following:

https://k8s-testgrid.appspot.com/release-1.6-upgrade-skew#gke-gci-1.5-gci-1.6-upgrade-cluster&width=20&exclude-non-failed-tests=

This test job is creating a 1.5 cluster, upgrading it to 1.6, then running 1.5 E2E tests against it.

The specific tests I need you to look at are:

[k8s.io] Daemon set [Serial] should run and stop complex daemon with node affinity
[k8s.io] SchedulerPredicates [Serial] validates that Inter-pod-Affinity is respected if not matching
[k8s.io] SchedulerPredicates [Serial] validates that InterPodAntiAffinity is respected if matching 2
[k8s.io] SchedulerPredicates [Serial] validates that NodeAffinity is respected if not matching
[k8s.io] SchedulerPredicates [Serial] validates that a pod with an invalid NodeAffinity is rejected
[k8s.io] SchedulerPredicates [Serial] validates that a pod with an invalid podAffinity is rejected because of the LabelSelectorRequirement is invalid
[k8s.io] SchedulerPredicates [Serial] validates that taints-tolerations is respected if not matching

**Comment by ethernetdan** [comment]
Opening until can be confirmed non-issue, @davidopp could you make a call on this one?

**Comment by davidopp** [comment]
These are almost definitely fine but I haven't had time to look.

**Comment by ethernetdan** [comment]
@davidopp would it be possible to check by EOD? I'd like to get ahead of problems if they exist

**Comment by davidopp** [comment]
Yes I will check by tonight.

**Comment by davidopp** [comment]
OK I checked and these are all expected failures.

The reason I was initially surprised by the failures is that we still support the alpha annotations for affinity in 1.6, so in theory those tests should work. But then I remembered that the alpha support is now flag-gated and this test suite doesn't enable the flag gate. I'm going to file an issue for us to set up a test suite that enables the flag and tests the annotation version of this feature, but that's not necessary for 1.6.
